### PR TITLE
LPS-124108 [DO NOT FORWARD UNTIL GA1 RELEASE] Add externalReferenceCode infrastructure to Document Library services

### PR DESCRIPTION
**Pull request by javierdearcos**

### CODE REVIEW

**Comment by javierdearcos** [comment]
I've followed the same approach that I did in MBMessage and was validated by Brian, that is add the externalReferenceCode as the first parameter in the addFile methods.

There was a lot of infrastructure were I have to add the parameter, DLAppService and DocumentRepository class hierarchy. Tried to do it in fewer places but I couldn't because of the way the hierarchy is implemented.

Take into account the export/import process and the upgrade process.

I've deprecated the old methods without externalReferenceCode as Brian requested for the ones in MBMessage services.

I've added tests to be sure about the new code and every existing test is working, but I don't know if it is a good idea to pass the pull through the Lima team in case I miss something that shouldn't be done

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java [hunk 1]
```diff
@@ -150,6 +152,138 @@ public FileEntry addFileEntry(
...
+	 * @param  changeLog the file's version change log
+	 * @param  file the file's data (optionally <code>null</code>)
+	 * @param  serviceContext the service context to be applied. Can set the
+	 *         asset category IDs, asset tag names, and expando bridge
+	 *         attributes for the file entry. In a Liferay repository, it may
+	 *         include:  <ul> <li> fileEntryTypeId - ID for a custom file entry
+	 *         type </li> <li> fieldsMap - mapping for fields associated with a
+	 *         custom file entry type </li> </ul>
+	 * @return the file entry
+	 * @throws PortalException if a portal exception occurred
```
Comment (nhpatt) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java]: I think you have to add an @review

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppLocalServiceImpl.java [hunk 2]
```diff
@@ -88,6 +88,12 @@
  */
 public class DLAppLocalServiceImpl extends DLAppLocalServiceBaseImpl {
 
+	/**
```
Comment (nhpatt) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppLocalServiceImpl.java:91]: I prefer to not deprecate the methods. Can you send it without it and see if it passes? Forcing to pass a external code to people only using the services, that won't ever use it and force them to know what it is...

Comment (javierdearcos) [reply portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppLocalServiceImpl.java:91]: Sure. I'll remove the commits deprecating the methods

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java (hunk 1 above)
Comment (javierdearcos) [reply portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java:189]: I've made a quick search and the first review javadoc annotation that I've found is from August 2019. Let's hope they review these comments a little bit quicker 😄

**Comment by javierdearcos** [comment]
I'm thinking that, anyway, maybe we should wait until the GA1 realease. I'll continue fixing the things you find and shelve the PR until then

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java (hunk 1 above)
Comment (nhpatt) [reply portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java:189]: 😂 I think documentation is swamped with liferay.learn and is ignoring those

**Comment by javierdearcos** [comment]
ci:test:sf

**Comment by javierdearcos** [comment]
ci:test:relevant

**Comment by javierdearcos** [comment]
ci:test:sf

**Comment by javierdearcos** [comment]
ci:test:relevant

**Comment by javierdearcos** [comment]
ci:stop

**Review by nhpatt — APPROVED** [review]

**Pull request closed without merging**